1311 orm adr (#235)
Energinet-DataHub/energy-origin: doc/architecture/adr/0020-PROPOSAL-authentication-orm.md

# Authentication ORM

* Status: Proposed
* Deciders: @MortMH, @Frydensberg1, @CodeReaper, @robertrosborg
* Date: 2023-02-06

---

## Context and Problem Statement

We need an ORM for the Auth domain to handle entity mapping and database migrations.

---

## Considered Options

* Entity Framework
* Dapper

---

## Decision Outcome

We have chosen Entity Framework.

## Rationale

Both Entity Framework and Dapper are capable ORMs, but we've chosen Entity Framework for the following main reasons:

> * Built-in migration handling and generation.
> * Improved productivity and faster development.
> * Well known and supported in .NET ecosystem.

Reviewer: ```suggestion
* Built-in migration handling and generation.
* Well known and supported in .NET ecosystem.
```

# Authentication ORM

* Status: Proposed
* Deciders: @MortMH, @Frydensberg1, @CodeReaper, @robertrosborg
* Date: 2023-02-06

---

## Context and Problem Statement

> MartenDB EventStore has previously been proposed as the database of choice, but [ADR-0015](0015-domains-responsible-for-persistance.md) allows each domain to choose the database that best suits it. For the Auth domain we've therefore decided to go for a simpler solution in the form of a relational PostgreSQL database.

Reviewer: ```suggestion
MartenDB EventStore has previously been proposed as the database of choice, but [ADR-0015](0015-domains-responsible-for-persistance.md) allows each domain to choose the database that best suits it. For the Auth domain we have therefore decided to go for a simpler solution in the form of a relational database. PostgreSQL has previously been the go-to choice and we found no reason to change this.
```

> Because of this we need an ORM to handle entity mapping and database migrations.

Reviewer: ```suggestion
We looked into finding an ORM to handle entity mapping and database migrations.
```

---

## Considered Options

* Entity Framework
* Dapper

---

## Decision Outcome

We have chosen Entity Framework.

## Rationale

> Both Entity Framework and Dapper are capable ORMs and a choice between them will often come down to personal preferences, but the main differentiating reasons we've chosen Entity Framework are:

Reviewer: ```suggestion
Both Entity Framework and Dapper are capable ORMs and a choice between them will often come down to personal preferences, but the main differentiating reasons we have chosen Entity Framework are:
```

> * Built-in migration handling and generation.
> * Improved productivity and faster development.
> * Well known and supported in .NET ecosystem.

Reviewer: ```suggestion
* Built-in migration handling and generation.
* Avoidance of manually writing SQL.
* Well known and supported in .NET ecosystem.
* Improved productivity and faster development.
```
